File: 1021. Remove Outermost Parentheses.java
// Medium
// Stack
// Time:O(n),Space:O(n)
// https://leetcode.cn/problems/remove-outermost-parentheses/

class Solution {
  public String removeOuterParentheses(String s) {
      StringBuilder sb = new StringBuilder();
      int depth = 0;
      for (char c : s.toCharArray()) {
          if (c == '(') {
              if (depth > 0) sb.append(c); // 非外层括号，保留
              depth++;
          } else {
              depth--;
              if (depth > 0) sb.append(c); // 非外层括号，保留
          }
      }
      return sb.toString();
  }
}

File: 1001-2000/1021. Remove Outermost Parentheses.py
# Easy
# String
# Time:O(n), Space:O(n)
# https://leetcode.cn/problems/remove-outermost-parentheses/

class Solution:
    def removeOuterParentheses(self, s: str) -> str:
        res = []
        depth = 0

        for c in s:
            if c == '(':
                if depth > 0:
                    res.append(c)
                depth += 1
            else:
                depth -= 1
                if depth > 0:
                    res.append(c)
        
        return ''.join(res)
    